File: app/src/main/java/com/fstyle/structure_android/utils/rx/SchedulerProvider.java
package com.fstyle.structure_android.utils.rx;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import rx.Scheduler;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by le.quang.dao on 28/03/2017.
 * Provides different types of schedulers.
 */

public final class SchedulerProvider implements BaseSchedulerProvider {
    @Nullable
    private static SchedulerProvider mInstance;

    // Prevent direct instantiation.
    private SchedulerProvider() {
    }

    public static synchronized SchedulerProvider getInstance() {
        if (mInstance == null) {
            mInstance = new SchedulerProvider();
        }
        return mInstance;
    }

    @Override
    @NonNull
    public Scheduler computation() {
        return Schedulers.computation();
    }

    @Override
    @NonNull
    public Scheduler io() {
        return Schedulers.io();
    }

    @Override
    @NonNull
    public Scheduler ui() {
        return AndroidSchedulers.mainThread();
    }
}

File: app/src/main/java/com/fstyle/structure_android/utils/rx/SchedulerProvider.kt
package com.fstyle.structure_android.utils.rx

import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Created by le.quang.dao on 28/03/2017.
 */

class SchedulerProvider private constructor() : BaseSchedulerProvider {

  override fun computation(): Scheduler {
    return Schedulers.computation()
  }

  override fun io(): Scheduler {
    return Schedulers.io()
  }

  override fun ui(): Scheduler {
    return AndroidSchedulers.mainThread()
  }

  private object SchedulerProvider {
    val instance = SchedulerProvider()
  }

  companion object {
    val instance: com.fstyle.structure_android.utils.rx.SchedulerProvider by lazy { SchedulerProvider.instance }
  }
}
